File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque

def solution(places):
    answer = []
    for place in places:
        answer.append(bfs(place))
    return answer

def bfs(place):
    starts = []
    for i in range(5):
        for j in range(5):
            if place[i][j] == 'P':
                starts.append([i, j])
    for start in starts:
        que = deque([start])
        visited = [[0 for col in range(5)] for row in range(5)]
        visited[start[0]][start[1]] = 1
        distance = [[0 for col in range(5)] for row in range(5)]
        while que:
            y, x = que.popleft()
            for i in range(4):
                dx = [-1, 1, 0, 0]
                dy = [0, 0, -1, 1]
                new_x = x + dx[i]
                new_y = y + dy[i]
                if 0<= new_x and new_x<5 and 0<= new_y and new_y<5 and visited[new_y][new_x] == 0:
                    if place[new_y][new_x] == 'O':
                        visited[new_y][new_x] = 1
                        distance[new_y][new_x] = distance[y][x] + 1
                        que.append([new_y,new_x])
                    if place[new_y][new_x] == 'P' and distance[y][x] <= 1:
                        return 0
    return 1

File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    public static int[] solution(String[][] places) {
        int[] answer = new int[5];
        for (int i = 0; i < 5; i++) {
            answer[i] = bfs(places[i]);
        }
        return answer;
    }
    public static int bfs(String[] place){
        List<int[]> starts = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (place[i].charAt(j) == 'P') {
                    starts.add(new int[] {i, j});
                }
            }
        }
        for (int[] start:starts) {
            Queue<int[]> queue = new LinkedList<>();
            queue.add(start);
            int[][] visited = new int[5][5];
            int[][] distance = new int[5][5];
            visited[start[0]][start[1]] = 1;
            while (!queue.isEmpty()) {
                int[] temp = queue.poll();
                int x = temp[1];
                int y = temp[0];
                int[] dx = {-1, 1, 0, 0};
                int[] dy = {0, 0, -1, 1};
                for (int i = 0; i < 4; i++) {
                    int newX = x + dx[i];
                    int newY = y + dy[i];
                    if (0<=newX && newX<=4 && 0<=newY && newY<=4 && visited[newY][newX] == 0) {
                        if (place[newY].charAt(newX) == 'O') {
                            visited[newY][newX] = 1;
                            distance[newY][newX] = distance[y][x] + 1;
                            queue.add(new int[] {newY, newX});
                        }
                        if (place[newY].charAt(newX) == 'P' && distance[y][x] <= 1) {
                            return 0;
                        }
                    }
                }
            }
        }

        return 1;
    }
}